File: dawn_cpp/compression.cpp
#include <algorithm>
#include <string>
#include <vector>

class BWT {
public:
    static std::pair<std::string, int> encode(const std::string& s) {
        int n = s.length();
        std::vector<std::string> rotations(n);
        for (int i = 0; i < n; i++) {
            rotations[i] = s.substr(i) + s.substr(0, i);
        }
        std::sort(rotations.begin(), rotations.end());
        std::string encoded(n, '\0');
        for (int i = 0; i < n; i++) {
            encoded[i] = rotations[i][n - 1];
        }
        int idx = std::find(rotations.begin(), rotations.end(), s) - rotations.begin();
        return {encoded, idx};
    }

    static std::string decode(const std::string& encoded, int idx) {
        int n = encoded.length();
        std::vector<std::string> sortedRotations(n);
        for (int i = 0; i < n; i++) {
            sortedRotations[i] = encoded.substr(i, 1) + lastChar(encoded, i);
        }
        std::sort(sortedRotations.begin(), sortedRotations.end());
        std::string decoded(n, '\0');
        for (int i = 0; i < n; i++) {
            decoded[i] = sortedRotations[i][n - 1];
        }
        return decoded.substr(idx) + decoded.substr(0, idx);
    }

private:
    static std::string lastChar(const std::string& s, int i) {
        std::string temp = s.substr(0, i) + s.substr(i + 1);
        std::sort(temp.begin(), temp.end());
        return std::string(1, temp[temp.length() - 1]);
    }
};

File: dawn_go/compression.go
package bwt

func encode(s string) (string, int) {
	n := len(s)
	rotations := make([]string, n)
	for i := range rotations {
		rotations[i] = s[i:] + s[:i]
	}
	sortedRotations := sortStringSlice(rotations)
	encoded := make([]byte, n)
	for i, r := range sortedRotations {
		encoded[i] = r[n-1]
	}
	idx := findIndex(s, sortedRotations)
	return string(encoded), idx
}

func decode(encoded string, idx int) string {
	n := len(encoded)
	sortedRotations := make([]string, n)
	for i := range sortedRotations {
		sortedRotations[i] = string(encoded[i]) + string(lastChar(encoded[:i]+string(encoded[i])))
	}
	sortedRotations = sortStringSlice(sortedRotations)
	decoded := make([]byte, n)
	for i, r := range sortedRotations {
		decoded[i] = r[n-1]
	}
	return string(decoded[idx:]) + string(decoded[:idx])
}

func lastChar(s string) byte {
	return sortStringSlice([]string{s})[0][len(s)-1]
}

func findIndex(s string, sortedRotations []string) int {
	for i, r := range sortedRotations {
		if r == s {
			return i
		}
	}
	return -1
}

func sortStringSlice(s []string) []string {
	sortedSlice := make([]string, len(s))
	copy(sortedSlice, s)
	sort.Strings(sortedSlice)
	return sortedSlice
}